discovery/openstack/openstack.go

```diff
@@ -0,0 +1,202 @@
+	port     int
+}
+
+// NewDiscovery returns a new OpenStackDiscovery which periodically refreshes its targets.
+func NewDiscovery(conf *config.OpenstackSDConfig) (*Discovery, error) {
+	opts := gophercloud.AuthOptions{
+		IdentityEndpoint: conf.IdentityEndpoint,
+		Username:         conf.Username,
+		UserID:           conf.UserID,
+		Password:         conf.Password,
+		TenantName:       conf.TenantName,
+		TenantID:         conf.TenantID,
+		DomainName:       conf.DomainName,
+		DomainID:         conf.DomainID,
+	}
+
+	return &Discovery{
+		os:       &opts,
+		region:   conf.Region,
+		interval: time.Duration(conf.RefreshInterval),
+		port:     conf.Port,
+	}, nil
+}
+
+// Run implements the TargetProvider interface.
+func (d *Discovery) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
+	ticker := time.NewTicker(d.interval)
+	defer ticker.Stop()
+
+	// Get an initial set right away.
+	tg, err := d.refresh()
+	if err != nil {
+		log.Error(err)
+	} else {
+		select {
+		case ch <- []*config.TargetGroup{tg}:
+		case <-ctx.Done():
+			return
+		}
+	}
+
+	for {
+		select {
+		case <-ticker.C:
+			tg, err := d.refresh()
+			if err != nil {
+				log.Error(err)
+				continue
+			}
+
+			select {
+			case ch <- []*config.TargetGroup{tg}:
+			case <-ctx.Done():
+				return
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (d *Discovery) refresh() (tg *config.TargetGroup, err error) {
+	t0 := time.Now()
+	defer func() {
+		openstackSDRefreshDuration.Observe(time.Since(t0).Seconds())
+		if err != nil {
+			openstackSDRefreshFailuresCount.Inc()
+		}
+	}()
+	provider, err := openstack.AuthenticatedClient(*d.os)
+
+	if err != nil {
+		return nil, fmt.Errorf("could not create openstack session: %s", err)
+	}
+	client, err := openstack.NewComputeV2(provider, gophercloud.EndpointOpts{
+		Region: d.region,
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("could not create openstack compute session: %s", err)
+	}
+
+	opts := servers.ListOpts{}
+	pager := servers.List(client, opts)
+
+	tg = &config.TargetGroup{}
+
+	err = pager.EachPage(func(page pagination.Page) (bool, error) {
+		serverList, err := servers.ExtractServers(page)
+		if err != nil {
+			return false, fmt.Errorf("could not extract servers: %s", err)
+		}
+
+		for _, s := range serverList {
+			labels := model.LabelSet{
+				openstackLabelInstanceID: model.LabelValue(s.ID),
+			}
+
+			for _, address := range s.Addresses {
+				md, ok := address.([]interface{})
+
+				if !ok || len(md) < 1 {
+					continue
+				}
+
+				md1, ok := md[0].(map[string]interface{})
+				if !ok {
+					continue
+				}
+
+				addr := net.JoinHostPort(md1["addr"].(string), fmt.Sprintf("%d", d.port))
+
+				labels[model.AddressLabel] = model.LabelValue(addr)
+				break
+			}
+
+			labels[openstackLabelInstanceState] = model.LabelValue(s.Status)
+			labels[openstackLabelInstanceName] = model.LabelValue(s.Name)
+			labels[openstackLabelInstanceFlavor] = model.LabelValue(s.Flavor["id"].(string))
```

Comment:
You probably want to do a checked type assertion here too, otherwise this might panic if it's not really a `string`.